Feature/m0127 cc 568 tutorial build micro ej firmware (#75)
MicroEJ/docs: Tutorials/tutorialUnderstandMicroEJFirmwareBuild.rst

Understand how to build a MicroEJ Firmware and its dependencies
===============================================================

A MicroEJ Firmware is built from several input resources and tools.
Each component has a specific version that must be carefully respected

> in order to build a firmware.

Author: It emphasizes that one has to respect versions and dependencies between components. You can't make a firmware from incompatible components. I'll try to rephrase it.

> in order to build a firmware.

Author: Done. Thoughts on the rephrasing?

> in order to build a firmware.

Reviewer: Good

This document describes the components and the process involved in the
build of a MicroEJ Firmware.

> A good knowledge of the :ref:`chapter-glossary` is required.

Reviewer: ```suggestion
Good knowledge of the :ref:`chapter-glossary` is required.
```

The Components
--------------

As depicted in the following image, several resources and tools are
used to build a MicroEJ Firmware.

> .. image:: images/qa_resources-v3.png

Reviewer: "MEJ32 Architecture" -> "MicroEJ Architecture"
"MEJ32 Repository" -> "MicroEJ Repository"
"MEJ32 Arch/Compiler" -> "MicroEJ Architecture"

Author: Do you know where is located the source of this image? I couldn't find it.

Author: I redid it. WDYT?

Reviewer: Good

	:align: center

MicroEJ Architecture (.xpf, .xpfp)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

A MicroEJ Architecture contains the runtime port to a target
instruction set (ISA) and a C compiler (CC) and MicroEJ Foundation
Libraries.

> The MicroEJ Architectures are distributed into 2 formats:

Reviewer: ```suggestion
The MicroEJ Architectures are distributed into two formats:
```

* EVAL: evaluation license with runtime limitations (explained in SDK
  developer guide).
* DEV: production license (only MicroEJ sales & Customer Care team distribute this version).

> The list of supported MicroEJ Architectures can be found here

Reviewer: ```suggestion
The supported MicroEJ Architectures are listed here
```

* For DEV license only: SDK license site https://license.microej.com/
  (:guilabel:`MyProduct` > :guilabel:`Download additional products`
  will list the downloads available).  See
  :ref:`gettingstarted-installlicensesdev` for help with DEV license.
* MicroEJ sales or customer care team if the requested architecture is not listed as available.

See :ref:`architecture_import` for a description on how to import a
MicroEJ Architecture.

MicroEJ Platform Source (.zip)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This package includes:

* a C Board Support Package (BSP) with or without RTOS and C drivers
* a MEJ32 Architecture
* the abstraction layers implementation of the platform architecture and foundations libraries
* the MicroEJ Simulator and its associated MicroEJ mocks

The platform .zip files contain:

* ``<platform>-configuration``: The configuration of the MicroEJ
  Platform
* ``<platform>-bsp``: The C code for the board-specific files
  (drivers).
* ``<platform>-fp``: Front panel mockup for the simulator.

See :ref:`new_platform_creation` to learn how to create a MicroEJ
Platform using a MicroEJ Platform Source project.

> Depending on the requirements of the project, the MicroEJ Platform can

Reviewer: ```suggestion
Depending on the project's requirements, the MicroEJ Platform can
```

> be connected in various ways to the BSP, see :ref:`bsp_connection` for

Reviewer: ```suggestion
be connected in various ways to the BSP; see :ref:`bsp_connection` for
```

more information on how to do it.

MicroEJ Application
~~~~~~~~~~~~~~~~~~~

A MicroEJ Application is a Java project that can be configured (in the
:guilabel:`Run configurations ...` properties):

1. to either run on:

  * a simulator (computer desktop),
  * a device (actual embedded hardware).

2. to setup:

  * memory (example: Java heap, Java stack),
  * foundation libraries,

>   * etc…

Reviewer: ```suggestion
  * etc.
```

To run on a device, the application is compiled and optimized for a
specific MicroEJ Platform.  It generates a ``microejapp.o`` (native

> object code) that is linked with the

Reviewer: ```suggestion
object code) linked with the
```

``<platform>-bsp`` project.

> To import an existing MicroEJ Application as zip project in the SDK:

Reviewer: ```suggestion
To import an existing MicroEJ Application as a zipped project in the SDK:
```

* Go to :guilabel:`File` > :guilabel:`Import…` > :guilabel:`General` >
  :guilabel:`Existing Projects into Workspace` > :guilabel:`Select
  archive file` > :guilabel:`Browse…`.
* Select the zip of the project (e.g. ``x.zip``).
* And select :guilabel:`Finish` import.

See :ref:`microej.sample.applications` for more information on how to

> create, configure and develop a MicroEJ Application.

Reviewer: ```suggestion
create, configure, and develop a MicroEJ Application.
```

C Toolchain (GCC, KEIL, IAR, …)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Used to compile and link the following files into the final firmware
(binary, hex, elf, … that will be programmed on the hardware):

  * the ``microejapp.o`` (application),
  * the ``microejruntime.lib`` or ``microejruntime.a`` (platform),
  * the BSP C files (drivers).

Module (Ivy) repository
~~~~~~~~~~~~~~~~~~~~~~~

A module (Ivy) repository provides the modules required to build
MicroEJ Platforms and MicroEJ Applications.

* The MicroEJ Central Repository is an online repository of software

>   artifacts (libraries, tools, …), see

Reviewer: ```suggestion
  artifacts (libraries, tools, etc.), see
```

>   https://repository.microej.com/. This repository can also be used as an offline repository, see https://developer.microej.com/central-repository/.

Reviewer: ```suggestion
  https://repository.microej.com/. One can also use this repository as an offline repository, see https://developer.microej.com/central-repository/.
```

Author: Not in favor of this. We use the passive voice throughout the document, not the active one.

Reviewer: OK. I tried to remove as much as possible the passive voice from the document.
The active voice can provide more clarity and brevity. Grammarly advises against using the passive voice.

* (Optional) It can be extended with an offline repository (``.zip``)

>   that can be imported in the workspace (see

Reviewer: ```suggestion
  and imported in the workspace (see
```

Author: Disagree. It's the offline repository that can be imported. Not the Central Repository to which the "It" refers to.

Reviewer: Yes

See :ref:`module_repository` for more information.

Dependencies Between Components
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* A MicroEJ Architecture targets a specific instruction set (ISA) and
  a specific C compiler (CC).

  * The C toolchain used for the MicroEJ Architecture must be the same
    as the one used to compile and link the BSP project and the
    MicroEJ Firmware.

* A MicroEJ Platform consists of the aggregation of both a MicroEJ
  Architecture and a BSP with a C toolchain.

  * Changing either the MicroEJ Architecture or the C toolchain
    results in a change of the MicroEJ Platform.

* A MicroEJ Application is independent of the MicroEJ Architecture.

  * It can run on any MicroEJ Platform as long the platform provides the required APIs.

  * To run a MicroEJ Application on a new device, create a new
    MicroEJ Platform for this device with the exact
    same features.  The MicroEJ Application will not require any change.

How to build
------------

The process to build a MicroEJ Firmware is two-fold:

1. Build a MicroEJ Platform
2. Build a MicroEJ Application into a MicroEJ Firmware

The MicroEJ Application is compiled against the MicroEJ Platform to

> produce the MicroEJ Firmware that will be deployed on the target

Reviewer: ```suggestion
produce the MicroEJ Firmware deployed on the target
```

device.

.. note::

   The MicroEJ Application runs onto
   the MicroEJ Simulator using the mocks provided by the MicroEJ
   Platform.

The next schema presents the steps to build a MicroEJ Mono-Sandbox
Firmware (previously known as MicroEJ Single-app Firmware).  The steps
are:

1. Build the MicroEJ Platform using MicroEJ SDK

2. Build the MicroEJ Application into a ``microejapp.o`` using MicroEJ SDK

3. Compile the BSP C sources into ``.o`` using the C toolchain

4. the BSP (``.o``) and the MicroEJ Application (``microejapp.o``) and
   the MicroEJ Platform (``microejruntime.a``) are linked by the C toolchain to produce a
   final ELF or binary called MicroEJ Firmware (e.g. ``application.out``).

> .. image:: images/build_microej_mono_sandbox_firmware_numbered.PNG

Reviewer: Use schema from https://documentation.cross/share/page/site/m0127customercare/document-details?nodeRef=workspace://SpacesStore/05857911-5344-4c8b-9d02-3b2a34562975

Author: Done. Thanks.

	:align: center

See :ref:`bsp_connection` for more information on how to connect the
MicroEJ Platform to the BSP.

Dependencies Between Processes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* Rebuild the MicroEJ Platform in the following cases:

  * When the MicroEJ Architecture (``.xpf``) changes.

  * When a MicroEJ Foundation Library (``.xpfp``) changes.

  * When a Foundation Library changes, either when

>     * The public API (``.java`` or ``.h``) change.

Reviewer: ```suggestion
    * The public API (``.java`` or ``.h``) changes.
```
